### DISCUSSION TITLE
Composite action security

### DISCUSSION THREAD

**Discussion by 38ri581oq480** [community/community#27091] [Actions]
<p>Hello everyone!<br>
First post here, I have created a very simple composite action (<a href="https://github.com/marketplace/actions/create-and-use-docker-ssh-context" rel="noopener nofollow ugc">https://github.com/marketplace/actions/create-and-use-docker-ssh-context</a>).<br>
I am using this action to deploy to remote Docker swarm cluster by using Docker context.<br>
I have seen a lot of other actions that has post-run steps with cleanup of secrets, docker logout, etc.</p>
<ol>
<li>Is it possible to add post-run steps to composite action?</li>
<li>Does my action leaves ssh-key in ssh-agent after run, and is it possible that my private ssh-key can be compromised by running this action on GitHub hosted runners?</li>
</ol>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="38ri581oq480" data-post="1" data-topic="184287">
<div class="title">
<div class="quote-controls"></div>
 38ri581oq480:</div>
<blockquote>
<p>Is it possible to add post-run steps to composite action?</p>
</blockquote>
</aside>
<p>I don’t see anything about that in the <a href="https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions">Metadata syntax</a> documentation. For Docker and Javascript actions, yes, but I see no such thing for composite actions.</p>
<aside class="quote no-group" data-username="38ri581oq480" data-post="1" data-topic="184287">
<div class="title">
<div class="quote-controls"></div>
 38ri581oq480:</div>
<blockquote>
<p>Does my action leaves ssh-key in ssh-agent after run, and is it possible that my private ssh-key can be compromised by running this action on GitHub hosted runners?</p>
</blockquote>
</aside>
<p>The key will remain in the agent and the agent isn’t stopped, so yes, it stays available on the machine. On a GitHub hosted runner the impact should be limited because the runner VM is destroyed at the end of the job, on a self-hosted runner the stray ssh-agent might be more troublesome. Theoretically you could limit the impact by limiting the lifetime of the key inside the SSH agent (see the <code>-t</code> option of <code>ssh-add</code>), but that doesn’t solve the “stray process” problem.</p>

**Comment by 38ri581oq480** [comment]
<p>Should I rewrite this action as Docker or Javascript one to add ssh cleanup steps or is it 100% safe to use as is, if it is only running with GitHub hosted runners?</p>

**Comment by airtower-luna** [comment]
<p>It’s up to you whether it’s <em>safe enough</em> for your use case and security requirements with the current constraints. What I will say is that I wouldn’t offer it in the Marketplace without a cleanup process that makes sure the started <code>ssh-agent</code> instance gets stopped.</p>

**Comment by 38ri581oq480** [comment]
<p>Hey there!<br>
I have added some cleanup steps to action (<a href="https://github.com/marketplace/actions/docker-run-command-ssh-context" rel="noopener nofollow ugc">https://github.com/marketplace/actions/docker-run-command-ssh-context</a>). Can this action now be considered 100% safe to run on GitHub hosted runners and self-hosted runners? :upside_down_face:</p>

**Comment by airtower-luna** [comment]
<p>I’m not going to pretend <em>anything</em> is 100% secure. I will point out the security flaws I still see though:</p>
<ul>
<li>
<p>The way <code>ssh-keyscan</code> is used. Yes, it avoids the “unknown host” warning, but only by accepting the host key <code>ssh-keyscan</code> sees without any verification at all. If someone intercepts connections between the runner and your system, they’ll likely intercept both. The action should expect an actually known key (possibly with an opt-in for insecure behavior).</p>
</li>
<li>
<p>The cleanup steps won’t run if one of the earlier steps fails.</p>
</li>
<li>
<p>When passing the SSH private key to the agent, do it through an environment variable. That way the key doesn’t become part of the command line, which is then visible on the runner.</p>
</li>
</ul>

**Comment by 38ri581oq480** [comment]
<ol>
<li>
<p>Will add <code>ssh-public-key</code> input, thanks!</p>
</li>
<li>
<p>It is not possible to <strong>always</strong> run some cleanup steps with composite actions, right? :worried:<br>
I think I will switch to docker based action, and maybe trap errors to always perform a cleanup.<br>
Or is there a way to pass env vars from docker based action <code>entrypoint</code> to <code>post-entrypoint</code>? Because <code>post-entrypoint</code> will run on fresh container without env vars from <code>entrypoint</code>, correct? :nerd_face:<br>
…Do I really need to kill <code>ssh-agent</code> and remove keys from <code>ssh-add</code> with Docker based action? Does it leaks state anywhere after <code>entrypoint</code> has been run?</p>
</li>
<li>
<p>Is this way <code>ssh-add - &lt;&lt;&lt; "${{ inputs.ssh_private_key }}"</code> of passing a SSH private key to the agent safer than using echo and piping?</p>
</li>
</ol>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="38ri581oq480" data-post="7" data-topic="184287">
<div class="title">
<div class="quote-controls"></div>
 38ri581oq480:</div>
<blockquote>
<p>Will add <code>ssh-public-key</code> input, thanks!</p>
</blockquote>
</aside>
<p>I’d name it <code>ssh-host-key</code> to make the purpose a little clearer, but good! :slightly_smiling_face:</p>
<aside class="quote no-group" data-username="38ri581oq480" data-post="7" data-topic="184287">
<div class="title">
<div class="quote-controls"></div>
 38ri581oq480:</div>
<blockquote>
<p>It is not possible to <strong>always</strong> run some cleanup steps with composite actions, right? :worried:</p>
</blockquote>
</aside>
<p>I’m afraid not, unless there is some undocumented <code>post</code> feature (and I assume there isn’t). :confused:</p>
<aside class="quote no-group" data-username="38ri581oq480" data-post="7" data-topic="184287">
<div class="title">
<div class="quote-controls"></div>
 38ri581oq480:</div>
<blockquote>
<p>…Do I really need to kill <code>ssh-agent</code> and remove keys from <code>ssh-add</code> with Docker based action? Does it leaks state anywhere after <code>entrypoint</code> has been run?</p>
</blockquote>
</aside>
<p>The <code>ssh-agent</code> will be stopped when your Docker container is stopped. Explicitly stopping the agent would still be good to ensure it can do its cleanup, but the process and its socket will definitely be gone with the container. And you can do your own error handling in whatever you run inside the container, so an error in one command doesn’t have to prevent anything else from running. :slightly_smiling_face:</p>
<aside class="quote no-group" data-username="38ri581oq480" data-post="7" data-topic="184287">
<div class="title">
<div class="quote-controls"></div>
 38ri581oq480:</div>
<blockquote>
<p>Is this way <code>ssh-add - &lt;&lt;&lt; "${{ inputs.ssh_private_key }}"</code> of passing a SSH private key to the agent safer than using echo and piping?</p>
</blockquote>
</aside>
<p>Not what I was thinking of, but it should work! The point is that the process table (what you get e.g. with <code>ps xau</code>) contains command line arguments, so it’s good to avoid anything sensitive in those. The <code>&lt;&lt;&lt;</code> “here string” is handled by Bash, so there’s no extra process where it might show up.</p>

**Comment by 38ri581oq480** [comment]
<p>… new day - new problems<br>
Tried to switch to Docker based action.<br>
PROs:</p>
<ul>
<li>if any of steps fails, all next steps will work</li>
</ul>
<p>CONs:</p>
<ul>
<li>I have to login/logout to Docker Hub inside container too, so need to pass even more inputs.</li>
<li>Spent around 2 hours figuring out how to add host key to known_hosts. Now using <code>/etc/ssh/ssh_known_hosts</code> file to put host key. Because <code>$HOME/.ssh/known_hosts</code> file is not used inside Docker based action.</li>
</ul>
<p>This is insane. ssh-agent + GitHub Actions is such a big problem, actually. :expressionless: :expressionless: :expressionless:</p>
<p>I am almost giving up and switching to using Portainer on all swarm nodes…to deploy app from DockerHub webhook. So sad story :frowning:</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="38ri581oq480" data-post="9" data-topic="184287">
<div class="title">
<div class="quote-controls"></div>
 38ri581oq480:</div>
<blockquote>
<p>Because <code>$HOME/.ssh/known_hosts</code> file is not used inside Docker based action.</p>
</blockquote>
</aside>
<p>This seems odd. It should be used, but the user (and thus home directory) might change inside the container. Also SSH ignores config files if they’re not owned by the user in question, or access rights are too wide.</p>
